### ISSUE TITLE
CSV Loader

### ISSUE THREAD

**Issue by diegotsutsumi** [cmu-phil/tetrad#1129]
Hi Guys, I may be doing something wrong but I can't load a CSV file, it pops the following error:

Infer demiliter for file: business.csv
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException: -61
	at edu.pitt.dbmi.data.reader.util.TextFileUtils.inferDelimiter(TextFileUtils.java:154)
	at edu.cmu.tetradapp.editor.LoadDataSettings.getInferredDelimiter(LoadDataSettings.java:905)
	at edu.cmu.tetradapp.editor.LoadDataSettings.basicSettings(LoadDataSettings.java:517)
	at edu.cmu.tetradapp.editor.LoadDataDialog.showDataLoaderDialog(LoadDataDialog.java:202)
	at edu.cmu.tetradapp.editor.LoadDataAction.actionPerformed(LoadDataAction.java:104)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.AbstractButton.doClick(AbstractButton.java:369)
	at java.desktop/javax.swing.plaf.basic.BasicMenuItemUI.doClick(BasicMenuItemUI.java:980)
	at java.desktop/javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(BasicMenuItemUI.java:1024)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6589)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6354)
	at java.desktop/java.awt.Container.processEvent(Container.java:2261)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4966)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2319)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4798)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4914)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4543)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4484)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2305)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4798)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue.access$600(EventQueue.java:97)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Infer demiliter for file: business.graphml
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException: -61
	at edu.pitt.dbmi.data.reader.util.TextFileUtils.inferDelimiter(TextFileUtils.java:154)
	at edu.cmu.tetradapp.editor.LoadDataSettings.getInferredDelimiter(LoadDataSettings.java:905)
	at edu.cmu.tetradapp.editor.LoadDataSettings.basicSettings(LoadDataSettings.java:517)
	at edu.cmu.tetradapp.editor.LoadDataDialog.showDataLoaderDialog(LoadDataDialog.java:202)
	at edu.cmu.tetradapp.editor.LoadDataAction.actionPerformed(LoadDataAction.java:104)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.AbstractButton.doClick(AbstractButton.java:369)
	at java.desktop/javax.swing.plaf.basic.BasicMenuItemUI.doClick(BasicMenuItemUI.java:980)
	at java.desktop/javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(BasicMenuItemUI.java:1024)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6589)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6354)
	at java.desktop/java.awt.Container.processEvent(Container.java:2261)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4966)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2319)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4798)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4914)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4543)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4484)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2305)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4798)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue.access$600(EventQueue.java:97)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

Does anyone knows what's happening?

**Comment by kvb2univpitt** [comment]
@diegotsutsumi What version of Java are you using?

**Comment by diegotsutsumi** [comment]
@kvb2univpitt 
```tsutsumi@tsutsumi-HP:~$ java --version
openjdk 11.0.3 2019-04-16
OpenJDK Runtime Environment (build 11.0.3+7-Ubuntu-1ubuntu218.04.1)
OpenJDK 64-Bit Server VM (build 11.0.3+7-Ubuntu-1ubuntu218.04.1, mixed mode, sharing)
```

**Comment by kvb2univpitt** [comment]
@diegotsutsumi Currently, Tetrad only supports Java 8.  Please try Java 8.

**Comment by diegotsutsumi** [comment]
Hi @kvb2univpitt , thanks for the answer! I have tried java 8 as follows:
```
tsutsumi@tsutsumi-HP:~/Downloads$ sudo java -version
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-8u212-b03-0ubuntu1.18.04.1-b03)
OpenJDK 64-Bit Server VM (build 25.212-b03, mixed mode)
tsutsumi@tsutsumi-HP:~/Downloads$ sudo java -jar tetrad-gui-6.6.0-launch.jar
```

And I have the same error

**Comment by diegotsutsumi** [comment]
Here is the full log file for the command `sudo java -jar tetrad-gui-6.6.0-launch.jar 1> tetrad.log 2>&1`
[tetrad.log](https://github.com/cmu-phil/tetrad/files/3229544/tetrad.log)

**Comment by kvb2univpitt** [comment]
@diegotsutsumi Based on the error messages, either you're using the wrong java version or you have permission issue.  Are you able to share your dataset?  I want to see if I could load it on my end.

**Comment by diegotsutsumi** [comment]
@kvb2univpitt  Sure, it's attached here (I had to renamed it to *.log because github doesn't accept .csv)
[business.csv.log](https://github.com/cmu-phil/tetrad/files/3229623/business.csv.log)

**Comment by kvb2univpitt** [comment]
@diegotsutsumi It seems like your data does not have the right character encoding.  I ran an analyzer and got the following message:
```
Unknown character 'Ó' of byte value of -61 on line 7 at 18th character.
Line: 1157336950978369;ÓPERA CONCEPT HALL;Performance & Event Venue;PLACE;14125;3579;True;False;False;4.7;206;AUTOMATIC;14;-25.405574270764,-49.325339523889
```
Your data must "UTF-8".

**Comment by diegotsutsumi** [comment]
Hi Kevin, sorry for the late response and I appreciate very much your analysis.

Could you confirm me if the expected format is "UTF-8" ? I checked with the linux command  `file` and it returns me that the encoding is utf-8, as follows:

```tsutsumi@tsutsumi-HP:~/Downloads$ file business.csv.log 
business.csv.log: UTF-8 Unicode text, with very long lines
```

**Comment by kvb2univpitt** [comment]
Hi @diegotsutsumi Sorry.  I meant it has to be **us-ascii** encoded.
Your data file shows this:
> file -b --mime-encoding business.csv
> utf-8

My data file shows this:
> a_data]$ file -b --mime-encoding sim02_continuous_20var_1kcase.txt
> us-ascii

Basically, the byte value of the char has to be **0<= char <= 255**.  See [Ascii Table](http://www.asciitable.com/).

To get your data to run as is, you will need to do some code editing and compiling to bypass the delimiter inference.  In `edu.cmu.tetradapp.editor.LoadDataSettings `of tetrad-gui module, replace the following line (line 517):
```java
char inferredDelimiter = getInferredDelimiter(files.get(0));
```
with this:
```java
char inferredDelimiter = ';';
```
Basically, you're just hard coding it.  You should be able to load your data, as shown below:
![data](https://user-images.githubusercontent.com/10729309/58765511-d6863980-8541-11e9-8897-4011d071b44e.png)

Don't forget to choose discrete as your data type.  Also, you might need to clean up your dataset a bit.  There are some weird blank characters in there.

**Comment by kvb2univpitt** [comment]
Fixed.

**Comment by diegotsutsumi** [comment]
Thanks for the fix and the feedback!

**Comment by kvb2univpitt** [comment]
@diegotsutsumi You are welcome.  Thank you for the feedback!
